PolyEngine/Core/Src/pe/core/storage/impl/IndexedStringManager.cpp

```diff
@@ -0,0 +1,103 @@
+#include <pe/core/storage/impl/IndexedStringManager.hpp>
+
+namespace pe::core::storage::impl
+{
+
+constexpr size_t TIME_TO_LIVE = 100;
+
+IndexedStringManager& IndexedStringManager::get()
+{
+	static IndexedStringManager instance;
+	return instance;
+}
+
+const IndexedStringEntry* IndexedStringManager::registerString(const char* str)
+{
+	IndexedStringEntry* ret = nullptr;
+
+	auto it = m_entries.find(str);
+	if (it == m_entries.end())
+	{
+		// @todo(muniu) There are two allocations happening here.
+		// Consider fixing it if the performace is affected by this.
+		auto entry = std::make_unique<IndexedStringEntry>(str);
+		ret = entry.get();
+		// We need to create a new string_view, which points to the string with the same lifetime as entry.
+		// Otherwise we could have some memory issues.
+		std::string_view strView(entry.get()->get().GetCStr());
+		m_entries.emplace(strView, std::move(entry));
+	}
+	else
+	{
+		ret = it->second.get();
+	}
+	
+	ret->incrementRefCount();
+	ret->resetRemovalTimePoint();
+	return ret;
+}
+
+void IndexedStringManager::unregisterString(const IndexedStringEntry* entry)
+{
+	entry->decrementRefCount();
+
+	if (entry->getRefCount() == 0)
+	{
+		if (m_ttlEnabled)
+			scheduleErase(entry);
+		else
+			erase(entry);
+	}
+}
+
+void IndexedStringManager::scheduleErase(const IndexedStringEntry* entry)
+{
+	size_t removalTimePoint = m_tickCount + TIME_TO_LIVE;
+	HEAVY_ASSERTE(entry->getRemovalTimePoint().value_or(0) <= removalTimePoint,
+		"Cannot set removal time point to a previous one than already set!");
+	entry->setRemovalTimePoint(removalTimePoint);
+	m_ttlEntries.emplace(TTLEntry{removalTimePoint, entry});
+}
+
+void IndexedStringManager::erase(const IndexedStringEntry* entry)
+{
+	auto it = m_entries.find(entry->get().GetCStr());
+	ASSERTE(it != m_entries.end(), "Couldn't find indexed string entry for given string!");
+	m_entries.erase(it);
+}
+
+void IndexedStringManager::setTTLMode(bool enabled)
+{
+	if(m_ttlEnabled && !enabled)
+	{
+		bool bResult = m_ttlEntries.empty();
```

Comment:
not sure what to do here i keep getting "discarding return value of function with 'nodiscard' attribute queue.empty()" warning which is then treated as error by compiler